ciao/cmd/add.go

```diff
@@ -0,0 +1,902 @@
+
+`,
+	Run: func(cmd *cobra.Command, args []string) {
+		validateAddCommandArgs(instanceFlags)
+
+		var server api.CreateServerRequest
+		var servers api.Servers
+
+		populateCreateServerRequest(instanceFlags, &server)
+
+		err := validateCreateServerRequest(server)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error validating create server request: %s\n", err)
+		}
+
+		servers, err = C.CreateInstances(server)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error creating instances: %s\n", err)
+		}
+
+		if C.Template != "" {
+			tfortools.OutputToTemplate(os.Stdout, "instance-add", C.Template,
+				&servers.Servers, nil)
+		}
+
+		if len(servers.Servers) < instanceFlags.instances {
+			fmt.Println("Some instances failed to start - check the event log for details.")
+		}
+
+		for _, server := range servers.Servers {
+			fmt.Printf("Created new (pending) instance: %s\n", server.ID)
+		}
+	},
+}
+
+var poolAddCmd = &cobra.Command{
+	Use:  "pool [NAME]",
+	Long: `Add a pool to the cluster.`,
+	Args: cobra.MinimumNArgs(1),
+	Run: func(cmd *cobra.Command, args []string) {
+		poolFlags.name = args[0]
+		if poolFlags.name == "" {
+			fmt.Println("Missing required pool NAME parameter")
+			return
+		}
+
+		/* Handle the create option so we can add new pools */
+		if poolFlags.new {
+			err := C.CreateExternalIPPool(poolFlags.name)
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "Error creating pool: %s\n", err)
+			}
+
+			fmt.Printf("Created new pool: %s\n", poolFlags.name)
+			return
+		}
+
+		if poolFlags.subnet != "" {
+			// verify it's a good address.
+			_, network, err := net.ParseCIDR(poolFlags.subnet)
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "Error validating address: %s\n", err.Error())
+			}
+
+			if ones, bits := network.Mask.Size(); bits-ones < 2 {
+				fmt.Println("Use address mode to add a single IP address")
+			}
+
+			err = C.AddExternalIPSubnet(poolFlags.name, network)
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "Error adding external IP subnet: %s\n", err)
+				return
+			}
+		} else if len(args) < 1 {
+			fmt.Println("Missing any addresses to add")
+			return
+		} else {
+			for _, addr := range args {
+				// verify it's a good address
+				IP := net.ParseIP(addr)
+				if IP == nil {
+					fmt.Println("Invalid IP address")
+					return
+				}
+			}
+			err := C.AddExternalIPAddresses(poolFlags.name, args)
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "Error adding external IP addresses", err)
+				return
+			}
+		}
+
+		fmt.Printf("Added new address to: %s\n", poolFlags.name)
+	},
+}
+
+var tenantAddCmd = &cobra.Command{
+	Use:  "tenant [NAME]",
+	Long: `Create a new tenant with the supplied flags`,
+	Run: func(cmd *cobra.Command, args []string) {
+		if !C.IsPrivileged() {
+			fmt.Fprintf(os.Stderr, "Creating tenants is only available for privileged users")
+		}
+
+		if tenantFlags.tenantID == "" {
+			fmt.Fprintf(os.Stderr, "Missing required tenantID")
+			return
+		}
```

Comment:
I think this could be changed if you changed the controller api - you'd have to output the new tenant ID so that an admin can then use that to generate a cert.

Reply:
I would recommend we leave the requirement the way it is for now, and file an issue with a request to change this to not require a tenant ID. That way we can keep controller out of the picture with this PR.